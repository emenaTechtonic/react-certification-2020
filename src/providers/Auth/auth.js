import React, { useState, useEffect, createContext, useCallback } from 'react'
import { AUTH_STORAGE_KEY, AUTH_STORAGE_USER } from '../../utils/constants'
import { storage } from '../../utils/storage'

export const AuthContext = createContext(null)

export const AuthProvider = ({ children }) => {
  const [authenticated, setAuthenticated] = useState(false)

  useEffect(() => {
    const lastAuthState = storage.get(AUTH_STORAGE_KEY)
    const isAuthenticated = Boolean(lastAuthState)

    setAuthenticated(isAuthenticated)
  }, [])

  const login = useCallback((user) => {
    setAuthenticated(true)
    storage.set(AUTH_STORAGE_KEY, true)
    storage.set(AUTH_STORAGE_USER, user)
  }, [])

  const logout = useCallback((user) => {
    setAuthenticated(false)
    storage.set(AUTH_STORAGE_KEY, false)
    storage.set(AUTH_STORAGE_USER, user)
  }, [])

  return (
    <AuthContext.Provider value={{ login, logout, authenticated }}>
      {children}
    </AuthContext.Provider>
  )
}

// export { useAuth }
export default AuthProvider
