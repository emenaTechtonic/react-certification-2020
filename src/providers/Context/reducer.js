export const ContextReducer = (state, action) => {
  switch (action.type) {
    case 'TOGGLE_THEME':
      return {
        ...state,
        theme: !state.theme,
      }
    case 'OPEN_MENU':
      return {
        ...state,
        menu: action.payload,
      }
    case 'GET_DATA':
      return {
        ...state,
        data: action.payload,
      }
    case 'SET_VIEW':
      return {
        ...state,
        view: action.payload,
      }
    case 'SET_TARGET':
      return {
        ...state,
        target: action.payload,
      }
    case 'SET_USER':
      return {
        ...state,
        user: action.payload,
      }
    case 'SET_FAVORITES':
      return {
        ...state,
        favorites: action.payload,
      }

    default:
      return state
  }
}
