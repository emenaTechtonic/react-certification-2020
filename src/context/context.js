import React, { useEffect, createContext, useReducer, memo } from 'react'
import { useFetch } from '../utils/hooks/useFetch'
import { ContextReducer } from './reducer'

export const Context = createContext(null)

export const initialState = {
  target: 'Wizeline',
  theme: false,
  menu: false,
  view: 'home',
  data: [],
}
export const ContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(ContextReducer, initialState)
  const { data, isLoading } = useFetch(state.target)

  useEffect(() => {
    if (data) dispatch({ type: 'GET_DATA', payload: data })
  }, [data])

  return isLoading ? (
    <div> Loading...</div>
  ) : (
    <Context.Provider value={{ state, dispatch }}> {children} </Context.Provider>
  )
}
export default memo(ContextProvider)
