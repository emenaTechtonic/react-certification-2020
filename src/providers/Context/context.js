import React, { useEffect, createContext, useCallback, useReducer, memo } from 'react'
import { useFetch } from '../../utils/hooks/useFetch'
import { storage } from '../../utils/storage'
import { isElementIn, removeElement } from '../../utils/fns'
import {
  AUTH_STORAGE_USER,
  AUTH_STORAGE_KEY,
  AUTH_STORAGE_FAVORITES,
} from '../../utils/constants'
import { ContextReducer } from './reducer'

export const Context = createContext(null)

export const initialState = {
  target: 'Wizeline',
  theme: false,
  menu: false,
  view: 'home',
  data: [],
  user: { id: 0, name: '', avatarUrl: '' },
  favorites: [],
}
export const ContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(ContextReducer, initialState)
  const { data } = useFetch(state.target)

  useEffect(() => {
    if (data) dispatch({ type: 'GET_DATA', payload: data })
  }, [data])
  useEffect(() => {
    const lastAuthState = storage.get(AUTH_STORAGE_KEY)
    const isAuthenticated = Boolean(lastAuthState)

    if (isAuthenticated) {
      const lastUser = storage.get(AUTH_STORAGE_USER)
      const userAuthenticated = Object(lastUser)

      dispatch({ type: 'SET_USER', payload: userAuthenticated })

      const lastFavorites = storage.get(AUTH_STORAGE_FAVORITES)
      if (lastFavorites) {
        const favorites = lastFavorites
        dispatch({ type: 'SET_FAVORITES', payload: favorites })
      }
    }
  }, [])

  const saveFavoriteCard = useCallback((card) => {
    let favorites = []
    const lastFavorites = storage.get(AUTH_STORAGE_FAVORITES)
    if (lastFavorites) {
      favorites = lastFavorites
    }
    if (card) {
      const res = isElementIn(card.id, favorites)
      if (!res) {
        favorites.push(card)
        dispatch({ type: 'SET_FAVORITES', payload: favorites })
        storage.set(AUTH_STORAGE_FAVORITES, favorites)
      }
    }
  }, [])
  const removeFavoriteCard = useCallback((card) => {
    let favorites = []
    const lastFavorites = storage.get(AUTH_STORAGE_FAVORITES)
    if (lastFavorites) {
      favorites = lastFavorites
    }
    if (card) {
      const newFavorites = removeElement(card.id, favorites)
      dispatch({ type: 'SET_FAVORITES', payload: newFavorites })
      storage.set(AUTH_STORAGE_FAVORITES, newFavorites)
    }
  }, [])

  // return isLoading ? (
  //   <div> Loading...</div>
  // ) : (
  return (
    <Context.Provider value={{ state, dispatch, saveFavoriteCard, removeFavoriteCard }}>
      {children}
    </Context.Provider>
  )
}
export default memo(ContextProvider)
