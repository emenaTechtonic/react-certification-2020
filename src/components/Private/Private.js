import React, { useRef, useContext, useEffect, useState } from 'react'
import { Context } from '../../providers/Context/context'
import Card from '../Card/Card'
import StyledPrivate from './Private.styled'
import StyledPrivateDetails from './PrivateDetails.styled'
import Viewdetailes from '../Viewdetailes/Viewdetailes'
import Videogrid from '../Videogrid/Videogrid'

const Private = React.memo(() => {
  const { state, dispatch } = useContext(Context)
  const [view, setView] = useState(null)
  const [arrCards, setArrCards] = useState([])
  const [selectedCard, setSelectedCard] = useState(null)
  const sectionRef = useRef(null)

  useEffect(() => {
    const defRows = []
    if (selectedCard) {
      defRows.push(
        <StyledPrivateDetails key="favoriteDetails">
          <Viewdetailes
            id={selectedCard.id}
            key={`details_${selectedCard.id}`}
            title={selectedCard.title}
            description={selectedCard.description}
            link={setSelectedCard.link}
            w={setArrCards.w}
            h={setArrCards.h}
          />
          {state.favorites.length > 1 ? (
            <Videogrid id={selectedCard.id} key={`grid_${selectedCard.id}`}>
              {arrCards}
            </Videogrid>
          ) : null}
        </StyledPrivateDetails>
      )

      setView(defRows)
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [selectedCard])

  useEffect(() => {
    const getFavorites = () => {
      const rows = []
      state.favorites.forEach((element) => {
        rows.push(
          <Card
            detailed
            id={element.id}
            key={element.id}
            title={element.title}
            description={element.description}
            link={element.link}
            w={element.w}
            h={element.h}
            customClickEvent={() => {
              setSelectedCard(element)
            }}
          />
        )
      })

      const defRows = (
        <StyledPrivate key="favorites" ref={sectionRef}>
          {rows}
        </StyledPrivate>
      )

      setView(defRows)
      setArrCards(rows)
      setSelectedCard(null)
    }

    getFavorites()
  }, [dispatch, state.favorites, state.view])

  return <>{view}</>
})

export default Private
