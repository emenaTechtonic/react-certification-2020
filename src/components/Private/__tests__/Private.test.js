import React from 'react'
import renderer from 'react-test-renderer'
import { Context } from '../../../providers/Context/context'
import { AuthContext } from '../../../providers/Auth/auth'
import Private from '../Private'

const state = {
  target: 'Wizeline',
  theme: false,
  menu: false,
  view: 'favorites',
}
const login = null
const logout = null
const authenticated = false

describe('Testing Private component', () => {
  test('Renders Private without crashing', async () => {
    const rendered = renderer.create(
      <AuthContext.Provider value={{ authenticated, login, logout }}>
        <Context.Provider value={{ state }}>
          <Private />
        </Context.Provider>
      </AuthContext.Provider>
    )
    expect(rendered).toMatchSnapshot()
  })
})
