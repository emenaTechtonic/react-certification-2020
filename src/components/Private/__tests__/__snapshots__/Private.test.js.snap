// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Private component Renders Private without crashing 1`] = `null`;
